### ISSUE TITLE
Update algorithm for assigning students seats in the CBTF.

### ISSUE THREAD

**Issue by firasm** [PrairieLearn/PrairieTest-feedback#37]
Feedback about seat assignments from a proctor (@marasolen) :

> In the CBTF, the seats are currently assigned randomly. Naturally,
this means that sometimes large parts of the rooms are empty while the
other parts are quite dense with multiple students sitting beside each
other. We’d like to make the algorithm distribute students a bit more
evenly but in a way that is simple and room-independent other than the
number of seats. We believe that Anthony wrote the original code for
seat assignment that’s used when students tap their cards on the card
readers so we were hoping someone could modify it for us.
> 
> Our idea is: each time a student checks in, randomly select a seat
that is not beside any other already-taken seat. Once there are no
more seats that meet that criterion, the remaining seats are assigned
at random.

**Comment by firasm** [comment]
Other considerations: 

- Different behaviour when the CBTF is full (algorithm probably doesn't matter?) vs. half full (to concentrate or distribute?) vs. almost empty (concentrate to make proctoring easier)

- @nwalters512 said:

> The naive thing to do to avoid needing to construct a graph of seats is to treat the seats as being in a single line and distribute evenly along that. Though we've gotten other feedback that in underloaded sessions, people would actually prefer for students to be clustered in one part of the room to make proctoring easier.

**Comment by nwalters512** [comment]
A hybrid approach: always cluster towards the lowest-numbered seats while still trying to maintain empty seats between students when possible. This would really only have a big impact at <50% utilization, but would still probably be better than purely random.

**Comment by firasm** [comment]
We were trying very hard to not to have our solution require manually specifying the order, but none of the (good) algorithms we came up with were "independent of room design".

Just a thought - if we _could_ specify an order, I thought the easiest place to do it would be when creating a location and adding seats. We can already paste in a list of seats- what if it was just a custom list and seats were filled in that order 🤔
<img width="812" alt="Screenshot 2023-10-29 at 9 46 18 PM" src="https://github.com/PrairieLearn/PrairieTest-feedback/assets/2507459/32f3d8e6-a965-4718-ba17-8e2d431f5268">

**Comment by nwalters512** [comment]
Since seats are identified by numbers, I figured that would already give us an ordering that should give us at least some information about physical layout. That is, that seat $N$ is at least possibly next to seat $N+1$, and seat $N$ is almost definitely not next to seat $N+2$. I guess that still leaves you with the question of exactly which seats are "preferred" for low-utilization sessions, but I think we could pretty safely just say that we prefer lower-numbered seats.

**Comment by firasm** [comment]
Yes! I was worried for a sec, but our CBTF room starts numbering from 1 to $N$ where seat 1 is closest to the door so that would work for us too (without any renumbering)

![room014_2](https://github.com/PrairieLearn/PrairieTest-feedback/assets/2507459/7a3f3e9e-af75-4635-a3ef-b7144742f65f)

**Comment by mwest1066** [comment]
Some notes from the discussion with @firasm today:

* The use case for every-other seating is to separate students taking the same exam (or maybe better to separate students in the same course, irrespective of which exam they are taking). This is especially important for UBC right now because they aren't running simultaneous exams so all students in the room are taking the same exam.
* The Illinois CBTF wants to tight-pack students in low-usage sessions, which works fine so long as they are taking different exams from each other.
* It would be fine to assume that the room is numbered so that low numbers should be used first.

**Comment by trombonekenny** [comment]
In a previous scheduler, I used a least distancing method algorithm to generate two grouping of sets of exams that are roughly equal in reservation count (example: Group A has 3 exams with a total of 30 students, Group B has 1 exam with 32 students) and then alternated seating A B A B A, etc. That was easy insurance that students didn't sit next to other students taking the same exam, as long as the groups were about equal.

That works OK if you assume congruent numbered seats are next to each other, but building some sort of adjacency graph of seats is even better.

**Comment by mwest1066** [comment]
After discussion with Jim today:
* We want the seat assignment order to be different to the seat number, so this will probably entail a separate "seating order" column (non-unique, integer) for the seats in a location. If two seats have the same seating order then we'll order them by seat number.
* We'll separate students by one seat if they are taking the same exam (or same course?) but otherwise we'll aim to tightly pack.
